File: src/main/java/org/example/expert/service/TodoService.java
//package org.example.expert.service;
//
//import lombok.RequiredArgsConstructor;
//import org.example.expert.client.WeatherClient;
//import org.example.expert.controller.user.dto.response.UserInfoRespDto;
//import org.example.expert.exception.InvalidRequestException;
//import org.example.expert.controller.todo.dto.request.TodoSaveRequest;
//import org.example.expert.controller.todo.dto.request.TodoListReqDto;
//import org.example.expert.controller.todo.dto.request.TodoSearchReqDto;
//import org.example.expert.controller.todo.dto.response.TodoListRespDto;
//import org.example.expert.controller.todo.dto.response.TodoResponse;
//import org.example.expert.controller.todo.dto.response.TodoSaveResponse;
//import org.example.expert.controller.todo.dto.response.TodoSearchRespDto;
//import org.example.expert.domain.todo.Todo;
//import org.example.expert.domain.todo.TodoRepository;
//import org.example.expert.domain.user.User;
//import org.springframework.data.domain.Page;
//import org.springframework.data.domain.PageRequest;
//import org.springframework.data.domain.Pageable;
//import org.springframework.data.domain.Sort;
//import org.springframework.stereotype.Service;
//import org.springframework.transaction.annotation.Transactional;
//
//import static org.example.expert.controller.todo.dto.response.TodoSearchRespDto.*;
//
//@Service
//@RequiredArgsConstructor
//@Transactional(readOnly = true)
//public class TodoService {
//
//    private final TodoRepository todoRepository;
//    private final WeatherClient weatherClient;
//
//    @Transactional
//    public TodoSaveResponse saveTodo(User user, TodoSaveRequest todoSaveRequest) {
//        String weather = weatherClient.getTodayWeather();
//
//        Todo newTodo = new Todo(
//                todoSaveRequest.getTitle(),
//                todoSaveRequest.getContents(),
//                weather,
//                user
//        );
//        Todo savedTodo = todoRepository.save(newTodo);
//
//        return new TodoSaveResponse(
//                savedTodo.getId(),
//                savedTodo.getTitle(),
//                savedTodo.getContents(),
//                weather,
//                new UserInfoRespDto(user)
//        );
//    }
//
//    public TodoListRespDto getTodos(TodoListReqDto todoListReqDto) {
//        Pageable pageable = PageRequest.of(todoListReqDto.getPage(), todoListReqDto.getSize(), Sort.by(Sort.Direction.DESC, "modifiedAt"));
//        Page<Todo> todoPageList = todoRepository.findTodosByFilter(todoListReqDto, pageable);
//        return new TodoListRespDto(todoPageList);
//    }
//
//    public TodoResponse getTodo(Long todoId) {
//        return todoRepository.findByIdWithUser(todoId)
//                .map(TodoResponse::new)
//                .orElseThrow(() -> new InvalidRequestException("Todo Not found"));
//    }
//
//    public Todo findByIdOrFail(Long todoId){
//        return todoRepository.findById(todoId)
//                .orElseThrow(() -> new InvalidRequestException("Todo Not Found"));
//    }
//
//    public TodoSearchRespDto searchTodos(TodoSearchReqDto todoSearchReqDto) {
//        Pageable pageable = PageRequest.of(todoSearchReqDto.getPage(), todoSearchReqDto.getSize(), Sort.by(Sort.Direction.DESC, "createdAt"));
//        Page<TodoRespDto> todoPageList = todoRepository.searchTodosByFilter(todoSearchReqDto, pageable);
//        return new TodoSearchRespDto(todoPageList);
//    }
//}

File: src/main/java/org/example/expert/service/TodoService.kt
package org.example.expert.service

import org.example.expert.client.WeatherClient
import org.example.expert.controller.todo.dto.request.TodoListReqDto
import org.example.expert.controller.todo.dto.request.TodoCreateReqDto
import org.example.expert.controller.todo.dto.request.TodoSearchReqDto
import org.example.expert.controller.todo.dto.response.TodoInfoRespDto
import org.example.expert.controller.todo.dto.response.TodoListRespDto
import org.example.expert.controller.todo.dto.response.TodoSaveRespDto
import org.example.expert.controller.todo.dto.response.TodoSearchRespDto
import org.example.expert.domain.todo.Todo
import org.example.expert.domain.todo.TodoRepository
import org.example.expert.domain.user.User
import org.example.expert.exception.CustomApiException
import org.example.expert.exception.ErrorCode
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class TodoService(
    private val todoRepository: TodoRepository,
    private val weatherClient: WeatherClient
) {
    @Transactional
    fun saveTodo(user: User, todoCreateReqDto: TodoCreateReqDto): TodoSaveRespDto = todoCreateReqDto
            .toEntity(user, weatherClient.getTodayWeather())
            .let { todoRepository.save(it) }
            .let { TodoSaveRespDto(it) }

    fun getTodoList(todoListReqDto: TodoListReqDto): TodoListRespDto {
        val pageable = PageRequest.of(todoListReqDto.page, todoListReqDto.size, Sort.by(Sort.Direction.DESC, "modifiedAt"))
        val todoPage = todoRepository.findTodosByFilter(todoListReqDto, pageable)
        return TodoListRespDto(todoPage)
    }

    fun getTodo(todoId: Long): TodoInfoRespDto =
        todoRepository.findByIdWithUser(todoId)
            ?.let { TodoInfoRespDto(it) }
            ?: throw CustomApiException(ErrorCode.TODO_NOT_FOUND)

    fun findByIdOrFail(todoId: Long): Todo = todoRepository
        .findById(todoId)
        .orElseThrow { CustomApiException(ErrorCode.TODO_NOT_FOUND) }

    fun searchTodoList(todoSearchReqDto: TodoSearchReqDto): TodoSearchRespDto{
        val pageable = PageRequest.of(todoSearchReqDto.page, todoSearchReqDto.size, Sort.by(Sort.Direction.DESC, "createdAt"))
        val todoPage = todoRepository.searchTodosByFilter(todoSearchReqDto, pageable)
        return TodoSearchRespDto(todoPage)
    }

}